dids/bearerdid.go

```diff
@@ -0,0 +1,87 @@
+package dids
+
+import (
+	"fmt"
+
+	"github.com/tbd54566975/web5-go/crypto"
+	"github.com/tbd54566975/web5-go/jwk"
+)
+
+// BearerDID is a composite type that combines a DID with a KeyManager containing keys
+// associated to the DID. Together, these two components form a BearerDID that can be used to
+// sign and verify data.
+type BearerDID struct {
+	DID
+	crypto.KeyManager
+}
+
+// ToKeys exports a BearerDID into a portable format that contains the DID's URI in addition to
+// every private key associated with a verification method.
+func (d *BearerDID) ToKeys() (PortableDID, error) {
+	exporter, ok := d.KeyManager.(crypto.KeyExporter)
+	if !ok {
+		return PortableDID{}, fmt.Errorf("key manager does not implement KeyExporter")
+	}
+
+	resolutionResult := Resolve(d.URI)
+	if resolutionResult.GetError() != "" {
+		return PortableDID{}, fmt.Errorf("failed to resolve DID: %s", resolutionResult.GetError())
+	}
+	portableDID := PortableDID{URI: d.URI}
+	keys := make([]VerificationMethodKeyPair, 0)
+
+	didDoc := resolutionResult.Document
+	for _, vm := range didDoc.VerificationMethod {
+		keyAlias, err := vm.PublicKeyJwk.ComputeThumbprint()
+		if err != nil {
+			continue
+		}
+
+		key, err := exporter.ExportKey(keyAlias)
+		if err != nil {
+			continue
+		}
+
+		keys = append(keys, VerificationMethodKeyPair{
+			PublicKeyJWK:  vm.PublicKeyJwk,
+			PrivateKeyJWK: key,
+		})
+	}
+
+	portableDID.VerificationMethod = keys
+
+	return portableDID, nil
+}
+
+// FromKeys imports a BearerDID from a portable format that contains the DID's URI in addition to
+// every private key associated with a verification method.
+func BearerDIDFromKeys(portableDID PortableDID) (BearerDID, error) {
+	didURI, err := ParseURI(portableDID.URI)
+	if err != nil {
+		return BearerDID{}, err
+	}
+
+	keyManager := crypto.NewLocalKeyManager()
+	for _, vm := range portableDID.VerificationMethod {
+		keyManager.ImportKey(vm.PrivateKeyJWK)
+	}
+
+	return BearerDID{
+		DID:        didURI,
+		KeyManager: keyManager,
+	}, nil
+}
+
+// PortableDID is a serializable BearerDID. VerificationMethod contains the private key
+// of each verification method that the BearerDID's key manager contains
+type PortableDID struct {
+	URI                string                      `json:"uri"`
+	VerificationMethod []VerificationMethodKeyPair `json:"verificationMethod"`
+}
+
+// VerificationMethodKeyPair is a public/private keypair associated to a
+// BearerDID's verification method. Used in PortableDID
+type VerificationMethodKeyPair struct {
+	PublicKeyJWK  jwk.JWK `json:"publicKeyJwk"`
+	PrivateKeyJWK jwk.JWK `json:"privateKeyJwk"`
+}
```

Comment:
@frankhinek and I chatted a bit last week about this functionality. Frank wrote up a [really great doc](https://hackmd.io/SydUZUxlTiCLVVVp1FJ6DA?view#DidMethodtoKeys) about some of this functionality, using the name `DidKeySet`. 

Wondering what your thoughts were on that naming!

Reply:
Nice! I'll go ahead and update Swift's to match!

Reply:
It may be wise to label `PortableDID` as `PortableBearerDID`... hmmmmm... 🤔 

I'm assuming `BearerDID` implies the usage of a secure enclave whereas `PortableDID` being serializable obv will make the private keys available for usage outside of a secure enclave 

The case against `PortableBearerDID` would be because there's a degree of ambiguity in the defensibility of whether or not pulling the private keys out of the secure enclave invalidates those keys as no longer being a "bearer" instrument; that is to say, `PortableDID`'s aren't bearer instruments... but then it's like, if private keys are _ever_ pulled out of the enclave then at that point, following the same logic, those keys are no longer bearer instruments even if the keys are subsequently used from _within_ a secure enclave. Sheesh information theory is difficult. 

I think as we have it now is fine